config/src/config/config_optimizer.rs

```diff
@@ -189,16 +189,33 @@ fn optimize_public_network_config(
         let local_network_config_yaml = &local_config_yaml["full_node_networks"][index];
 
         // Only add seeds to testnet and mainnet (as they are long living networks)
-        if fullnode_network_config.network_id == NetworkId::Public
-            && local_network_config_yaml["seeds"].is_null()
-        {
-            if let Some(chain_id) = chain_id {
-                if chain_id.is_testnet() {
-                    fullnode_network_config.seeds = create_seed_peers(TESTNET_SEED_PEERS.into())?;
-                    modified_config = true;
-                } else if chain_id.is_mainnet() {
-                    fullnode_network_config.seeds = create_seed_peers(MAINNET_SEED_PEERS.into())?;
-                    modified_config = true;
+        if fullnode_network_config.network_id == NetworkId::Public {
+            if local_network_config_yaml["seeds"].is_null() {
+                if let Some(chain_id) = chain_id {
+                    if chain_id.is_testnet() {
+                        fullnode_network_config.seeds =
+                            create_seed_peers(TESTNET_SEED_PEERS.into())?;
+                        modified_config = true;
+                    } else if chain_id.is_mainnet() {
+                        fullnode_network_config.seeds =
+                            create_seed_peers(MAINNET_SEED_PEERS.into())?;
+                        modified_config = true;
+                    }
+                }
+            }
+
+            // The current identity is auto-generated. Load the identity from the filesystem or save to a new file
+            if let Identity::FromConfig(IdentityFromConfig {
+                source: IdentitySource::AutoGenerated,
+                key: config_key,
+                ..
+            }) = &fullnode_network_config.identity
+            {
+                let path = node_config.storage.dir().join("identity_key");
+                if let Some(loaded_identity) = Identity::load_identity(&path)? {
+                    fullnode_network_config.identity = loaded_identity;
```

Comment:
Hmm... the downside of this approach is that we're overwriting things twice, i.e., if the user provides `None` as the identity in the config file, the code will generate a random identity one (in `prepare_identity`), and then come here and overwrite it again. This could cause some confusion (e.g., if someone reads the identity value between the first generation, and the second overwrite).

Is it not possible to leave the config value as `None` in `prepare_identity`, and then when we get here, we decide what to do? This way there's only a single location to check for the logic, i.e., in this function, we could essentially just use the `prepare_identity` logic if the previous identity key is not found. Otherwise, we load it here?

What do you think?

Reply:
This makes sense. I moved all the identity generation stuff to the config optimizer